File: app/src/main/java/app/resketchware/ui/adapters/TonesAdapter.java
package app.resketchware.ui.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import app.resketchware.R;
import app.resketchware.ui.viewholders.TonesViewHolder;
import app.resketchware.utils.ColorPickerUtil;

import java.util.List;

public class TonesAdapter extends RecyclerView.Adapter<TonesViewHolder> {

    private List<Integer> colors = ColorPickerUtil.colorsMap.get(0);
    private int selectedColor = RecyclerView.NO_POSITION;
    private int selectedPalette = 0;
    private ToneClickListener toneClickListener;

    public interface ToneClickListener {
        void invoke(int color);
    }

    public void swapColors(int selection) {
        resetColor();
        selectedPalette = selection;
        colors = ColorPickerUtil.colorsMap.get(selection);
        notifyDataSetChanged();
    }

    @Override
    public TonesViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).
                inflate(R.layout.layout_tone_item, parent, false);
        return new TonesViewHolder(itemView);
    }

    @Override
    public int getItemCount() {
        return colors.size();
    }

    @Override
    public void onBindViewHolder(TonesViewHolder holder, int position) {
        holder.bind(colors.get(holder.getAbsoluteAdapterPosition()), selectedColor);
        holder.itemView.setOnClickListener(v -> {
            if (holder.getAbsoluteAdapterPosition() != selectedColor) {
                notifyItemChanged(selectedColor);
                selectedColor = holder.getAbsoluteAdapterPosition();
                notifyItemChanged(selectedColor);
                if (toneClickListener != null) {
                    toneClickListener.invoke(colors.get(selectedColor));
                }
            }
        });
    }

    public void setOnToneClickListener(ToneClickListener listener) {
        toneClickListener = listener;
    }

    public void resetColor() {
        selectedColor = RecyclerView.NO_POSITION;
    }
}

File: app/src/main/java/app/resketchware/ui/adapters/TonesAdapter.kt
package app.resketchware.ui.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import app.resketchware.R
import app.resketchware.ui.viewholders.TonesViewHolder
import app.resketchware.utils.ColorPickerUtil

class TonesAdapter : RecyclerView.Adapter<TonesViewHolder>() {

    private var colors = ColorPickerUtil.colorsMap.getValue(0)
    private var selectedColor = RecyclerView.NO_POSITION
    private var selectedPalette = 0
    var onToneClick: ((Int) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TonesViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.layout_tone_item, parent, false)
        return TonesViewHolder(itemView)
    }

    override fun getItemCount(): Int {
        return colors.size
    }

    override fun onBindViewHolder(holder: TonesViewHolder, position: Int) {
        holder.bind(colors[holder.absoluteAdapterPosition].toInt(), selectedColor)
        holder.itemView.setOnClickListener {
            if (holder.absoluteAdapterPosition != selectedColor) {
                notifyItemChanged(selectedColor)
                selectedColor = holder.absoluteAdapterPosition
                notifyItemChanged(selectedColor)
                onToneClick?.invoke(colors[holder.absoluteAdapterPosition].toInt())
            }
        }
    }

    fun swapColors(selection: Int) {
        resetColor()
        selectedPalette = selection
        colors = ColorPickerUtil.colorsMap.getValue(selection)
        notifyDataSetChanged()
    }

    fun resetColor() {
        selectedColor = RecyclerView.NO_POSITION
    }
}